File: api/services/tasks.py
from typing import List, Optional
from databridge.tasks_databridge import TasksDatabridge
from models.tasks import Task, TaskWithUser


class TasksService:
    def __init__(self, databridge: TasksDatabridge):
        self.databridge = databridge

    async def get_all_tasks(self) -> List[TaskWithUser]:
        """Get all tasks from all users"""
        tasks_data = await self.databridge.get_all_tasks()
        return [TaskWithUser(**task) for task in tasks_data]

    async def get_user_tasks(self, user_id: int) -> List[Task]:
        """Get all tasks for a specific user"""
        tasks_data = await self.databridge.get_user_tasks(user_id)
        return [Task(**task) for task in tasks_data]

    async def create_task(self, user_id: int, text: str) -> Optional[Task]:
        """Create a new task"""
        task_data = await self.databridge.create_task(user_id, text)
        if task_data:
            return Task(**task_data)
        return None

    async def toggle_task(self, task_id: int, user_id: int) -> Optional[Task]:
        """Toggle task completion status"""
        task_data = await self.databridge.toggle_task(task_id, user_id)
        if task_data:
            return Task(**task_data)
        return None

    async def delete_task(self, task_id: int, user_id: int) -> bool:
        """Delete a task"""
        return await self.databridge.delete_task(task_id, user_id)

File: app/services/tasks.ts
import { API_BASE_URL } from '@/constants/config';

export interface Task {
  id: number;
  user_id: number;
  text: string;
  completed: boolean;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
}

export interface TaskWithUser extends Task {
  user_name?: string | null;
  user_email?: string | null;
}

export interface CreateTaskRequest {
  text: string;
}

export const tasksService = {
  async getAllTasks(token: string): Promise<TaskWithUser[]> {
    const response = await fetch(`${API_BASE_URL}/tasks/all`, {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const errorText = await response.text().catch(() => 'Unknown error');
      throw new Error(`Failed to fetch tasks: ${response.status} - ${errorText}`);
    }

    return response.json();
  },

  async getMyTasks(token: string): Promise<Task[]> {
    const response = await fetch(`${API_BASE_URL}/tasks/my`, {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const errorText = await response.text().catch(() => 'Unknown error');
      throw new Error(`Failed to fetch my tasks: ${response.status} - ${errorText}`);
    }

    return response.json();
  },

  async createTask(token: string, text: string): Promise<Task> {
    const response = await fetch(`${API_BASE_URL}/tasks/`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify({ text }),
    });

    if (!response.ok) {
      const errorText = await response.text().catch(() => 'Unknown error');
      throw new Error(`Failed to create task: ${response.status} - ${errorText}`);
    }

    return response.json();
  },

  async toggleTask(token: string, taskId: number): Promise<Task> {
    const response = await fetch(`${API_BASE_URL}/tasks/${taskId}/toggle`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const errorText = await response.text().catch(() => 'Unknown error');
      throw new Error(`Failed to toggle task: ${response.status} - ${errorText}`);
    }

    return response.json();
  },

  async deleteTask(token: string, taskId: number): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/tasks/${taskId}`, {
      method: 'DELETE',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    });

    if (!response.ok) {
      const errorText = await response.text().catch(() => 'Unknown error');
      throw new Error(`Failed to delete task: ${response.status} - ${errorText}`);
    }
  },
};

